### PR TITLE
feat: allow builders to accept rest params and arrays

### PR DESCRIPTION
**Pull request by ImRodry**
**Please describe the changes this PR makes and why it should be merged:**
This PR changes all builders methods that previously only accepted arrays as input to accept both arrays and rest params. This was done by partially reverting #7759 and adding util functions to make it easy to support these two forms of input. There is no significant performance hit as this is not using Array.prototype.flat like before, it instead uses Array.isArray to check the kind of input that was provided.

**Status and versioning classification:**
- I know how to update typings and have done so, or typings don't need updating
- Code changes have been tested against the Discord API, or there are no code changes

### CODE REVIEW

File: packages/builders/src/components/ActionRow.ts [hunk 1]
```diff
@@ -38,17 +39,17 @@ export class ActionRowBuilder<T extends AnyComponentBuilder> extends ComponentBu
...
 	}
 
 	/**
 	 * Sets the components in this action row
 	 * @param components The components to set this row to
 	 */
-	public setComponents(components: T[]) {
-		this.components.splice(0, this.components.length, ...components);
+	public setComponents(...components: RestOrArray<T>) {
+		this.components.splice(0, this.components.length, ...normalizeArray(...components));
```
Comment (imranbarbhuiya) [inline packages/builders/src/components/ActionRow.ts:52]: Instead of spreading here again you can make normalizeArray to accept array input.

**Review by imranbarbhuiya — APPROVED** [review]

File: packages/builders/src/components/ActionRow.ts (hunk 1 above)
Comment (ImRodry) [reply packages/builders/src/components/ActionRow.ts:52]: It does I think? I just couldn’t figure out a way to make TS happy with this without doing this but if you know how then please suggest on the function itself

File: packages/builders/src/util/normalizeArray.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export function normalizeArray<T>(...arr: T[] | T[][]): T[] {
+	if (Array.isArray(arr[0])) return [...arr[0]];
+	return [...arr] as T[];
+}
```
Comment (imranbarbhuiya) [inline packages/builders/src/util/normalizeArray.ts:4]:
```suggestion
export function normalizeArray<T>(arr: RestOrArray<T>): T[] {
	if (Array.isArray(arr[0])) return arr[0]
	return arr as T[];
}
```

and remove spreading from all places
```diff
- fields = normalizeArray(...fields);
+ fields = normalizeArray(fields);
```

**Review by imranbarbhuiya — CHANGES_REQUESTED** [review]

File: packages/builders/src/messages/embed/Embed.ts [hunk 3]
```diff
@@ -12,17 +12,19 @@ import {
 	validateFieldLength,
 } from './Assertions';
 import { EmbedAuthorOptions, EmbedFooterOptions, RGBTuple, UnsafeEmbedBuilder } from './UnsafeEmbed';
+import { normalizeArray, RestOrArray } from '../../util/normalizeArray';
```
Comment (imranbarbhuiya) [inline packages/builders/src/messages/embed/Embed.ts:15]:
```suggestion
import { normalizeArray, type RestOrArray } from '../../util/normalizeArray';
```

**Review by imranbarbhuiya — CHANGES_REQUESTED** [review]

File: packages/builders/src/util/normalizeArray.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+export function normalizeArray<T>(arr: RestOrArray<T>): T[] {
+	if (Array.isArray(arr[0])) return arr[0];
+	return arr as T[];
+}
+
+export type RestOrArray<T> = T[] | T[][];
```
Comment (KhafraDev) [inline packages/builders/src/util/normalizeArray.ts:6]:
```suggestion
export type RestOrArray<T> = T[] | [T[]];
```

As you can see [here](https://www.typescriptlang.org/play?#code/C4TwDgpgBAShDOwDyAnAgilBDEAeAKgHxQC8U+A2gLpQA+UFlVVAsAFDsAmEAxgDZYU0AGYBXAHY9gASwD24qFk6cAwrIC2YeRHHB4ACgB0xnhq3idegFywEydJhy5xo9QCMIKQgEobEgNbisgDu4uzsSqpm2roGAIwANFAATEkAzN4RymqaMXr6FIkp6VSZbJE55pYGhUmpUGlUSbXFDaXhbKCQtoioGNggyQTEZEx05NTUHdz8giISUnIKFdEWsclGJqvVNnC9DgNDLu6ePn7igSFhHOXZ2+v6RfUZWVG5a3obLc-tt29VD2+JWaTxK3iAA), the third example will only fail with the updated type.

File: packages/builders/src/messages/embed/UnsafeEmbed.ts [hunk 5]
```diff
@@ -67,7 +69,8 @@ export class UnsafeEmbedBuilder {
 	 * Sets the embed's fields (max 25).
 	 * @param fields The fields to set
 	 */
-	public setFields(fields: APIEmbedField[]) {
+	public setFields(...fields: RestOrArray<APIEmbedField>) {
+		fields = normalizeArray(fields);
 		this.spliceFields(0, this.data.fields?.length ?? 0, ...fields);
```
Comment (NotSugden) [inline packages/builders/src/messages/embed/UnsafeEmbed.ts:74]:
```suggestion
		this.spliceFields(0, this.data.fields?.length ?? 0, ...normalizeArray(fields));
```

File: packages/builders/src/interactions/modals/UnsafeModal.ts [hunk 6]
```diff
@@ -57,7 +58,8 @@ export class UnsafeModalBuilder implements JSONEncodable<APIModalInteractionResp
 	 * Sets the components in this modal
 	 * @param components The components to set this modal to
 	 */
-	public setComponents(components: ActionRowBuilder<ModalActionRowComponentBuilder>[]) {
+	public setComponents(...components: RestOrArray<ActionRowBuilder<ModalActionRowComponentBuilder>>) {
+		components = normalizeArray(components);
 		this.components.splice(0, this.components.length, ...components);
```
Comment (NotSugden) [inline packages/builders/src/interactions/modals/UnsafeModal.ts:63]:
```suggestion
		components = ;
		this.components.splice(0, this.components.length, ...normalizeArray(components));
```

File: packages/builders/src/components/selectMenu/UnsafeSelectMenu.ts [hunk 7]
```diff
@@ -80,7 +82,8 @@ export class UnsafeSelectMenuBuilder extends ComponentBuilder<APISelectMenuCompo
 	 * Sets the options on this select menu
 	 * @param options The options to set on this select menu
 	 */
-	public setOptions(options: (UnsafeSelectMenuOptionBuilder | APISelectMenuOption)[]) {
+	public setOptions(...options: RestOrArray<UnsafeSelectMenuOptionBuilder | APISelectMenuOption>) {
+		options = normalizeArray(options);
```
Comment (NotSugden) [inline packages/builders/src/components/selectMenu/UnsafeSelectMenu.ts:86]: similarly here

File: packages/builders/src/components/selectMenu/UnsafeSelectMenu.ts [hunk 8]
```diff
@@ -67,7 +68,8 @@ export class UnsafeSelectMenuBuilder extends ComponentBuilder<APISelectMenuCompo
 	 * @param options The options to add to this select menu
 	 * @returns
 	 */
-	public addOptions(options: (UnsafeSelectMenuOptionBuilder | APISelectMenuOption)[]) {
+	public addOptions(...options: RestOrArray<UnsafeSelectMenuOptionBuilder | APISelectMenuOption>) {
+		options = normalizeArray(options);
```
Comment (NotSugden) [inline packages/builders/src/components/selectMenu/UnsafeSelectMenu.ts:72]: and here

**Review by NotSugden — CHANGES_REQUESTED** [review]

**Comment by ImRodry** [comment]
Rebased

**Review by kyranet — APPROVED** [review]

**Comment by BenPatersonxx** [comment]
@vladfrangu @iCrawl @SpaceEEC please review

**Review by SpaceEEC — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Pull request merged**